core/src/main/java/org/apache/accumulo/core/file/rfile/bcfile/Compression.java

```diff
@@ -497,69 +513,66 @@ public CompressionCodec load(Entry<Algorithm,Integer> key) {
     public abstract boolean isSupported();
 
     public Compressor getCompressor() throws IOException {
-      CompressionCodec codec = getCodec();
-      if (codec != null) {
-        Compressor compressor = CodecPool.getCompressor(codec);
-        if (compressor != null) {
-          if (compressor.finished()) {
-            // Somebody returns the compressor to CodecPool but is still using
-            // it.
-            LOG.warn("Compressor obtained from CodecPool already finished()");
-          } else {
-            LOG.debug("Got a compressor: " + compressor.hashCode());
-          }
-          /**
-           * Following statement is necessary to get around bugs in 0.18 where a compressor is referenced after returned back to the codec pool.
-           */
-          compressor.reset();
-        }
-        return compressor;
-      }
-      return null;
+      return compressorFactory.getCompressor(this);
     }
 
     public void returnCompressor(Compressor compressor) {
-      if (compressor != null) {
-        LOG.debug("Return a compressor: " + compressor.hashCode());
-        CodecPool.returnCompressor(compressor);
-      }
+      compressorFactory.releaseCompressor(this, compressor);
     }
 
     public Decompressor getDecompressor() throws IOException {
-      CompressionCodec codec = getCodec();
-      if (codec != null) {
-        Decompressor decompressor = CodecPool.getDecompressor(codec);
-        if (decompressor != null) {
-          if (decompressor.finished()) {
-            // Somebody returns the decompressor to CodecPool but is still using
-            // it.
-            LOG.warn("Decompressor obtained from CodecPool already finished()");
-          } else {
-            LOG.debug("Got a decompressor: " + decompressor.hashCode());
-          }
-          /**
-           * Following statement is necessary to get around bugs in 0.18 where a decompressor is referenced after returned back to the codec pool.
-           */
-          decompressor.reset();
-        }
-        return decompressor;
-      }
-
-      return null;
+      return compressorFactory.getDecompressor(this);
     }
 
     public void returnDecompressor(Decompressor decompressor) {
-      if (decompressor != null) {
-        LOG.debug("Returned a decompressor: " + decompressor.hashCode());
-        CodecPool.returnDecompressor(decompressor);
-      }
+      compressorFactory.releaseDecompressor(this, decompressor);
     }
 
     public String getName() {
       return compressName;
     }
   }
 
+  /**
+   * Default implementation will create new compressors.
+   */
+  private static CompressorFactory compressorFactory = new CompressorFactory(null);
```

Comment:
I think making this a `final AtomicReference<CompressorFactory>` would be good (corresponding comment below).